File: v2/plugins/core/delivery_retry.go
package core

// delivery_retry.go — gravar local é ÚLTIMO RECURSO, não primeiro reflexo.
//
// Junior Tip [a regra do dono, 2026-07-31]: "só pode gravar local caso o banco
// não responda depois de 3 tentativas; assim que voltar, descarregar tudo que tem
// local e limpar o local pra não ter risco". Antes desta regra, UMA falha
// qualquer — um blip de rede, uma eleição de líder de 200 ms — já criava estado
// local. Estado local é uma segunda cópia da verdade: enquanto ele existe, a
// memória do usuário está em dois lugares e só um deles é o AnhurDB. A fila tem
// de ser um túnel, não um armário.
//
// Junior Tip [por que 3 tentativas e NÃO 3 tentativas em tudo]: o dono disse "o
// banco não responde". Não responder é uma coisa; responder "não" é outra. Um 401
// (chave errada) ou um 409 (sessão cheia) são o servidor RESPONDENDO, e a
// resposta não muda na segunda tentativa — repetir só gastaria o tempo de um hook
// para chegar ao mesmo lugar. Então: 3 tentativas é o TETO para o que pode
// melhorar sozinho (timeout, conexão recusada, 5xx, 429, 408), e o que já foi
// decidido vai direto para a fila. O chunk é enfileirado nos dois casos — a regra
// muda QUANDO se grava local, nunca SE a memória é preservada.

import (
	"context"
	"errors"
	"time"

	"github.com/Yoven/AnhurDB-SDK/v2/golang/v2/client"
)

// maxDeliveryAttempts é o número do dono: três tentativas antes de aceitar o disco.
const maxDeliveryAttempts = 3

// deliveryRetryBackoff é a espera ANTES de cada tentativa (a primeira é imediata).
//
// Junior Tip [por que tão curto]: isto roda dentro de um hook que segura o fim do
// turno do usuário. O alvo é o blip — eleição de líder, reconexão, um 503 de
// deploy —, não uma queda longa. Queda longa é justamente o que a fila resolve, e
// a fila não tem pressa; o hook tem.
var deliveryRetryBackoff = []time.Duration{0, 300 * time.Millisecond, 1200 * time.Millisecond}

// deliveryOutcome descreve como uma entrega terminou. Existe para o chamador poder
// LOGAR a diferença entre "o banco não respondeu 3 vezes" e "o banco recusou" —
// eram indistinguíveis no log antigo, e foi por isso que uma rejeição permanente
// (HTTP 409) passou 10 dias parecendo uma queda temporária.
type deliveryOutcome struct {
	err          error
	attemptsMade int
	terminal     bool // o servidor respondeu e a resposta não vai mudar sozinha
}

// deliverWithRetry executa attemptDelivery até maxDeliveryAttempts vezes, parando
// cedo quando o erro é terminal ou o contexto foi cancelado.
func deliverWithRetry(ctx context.Context, attemptDelivery func() error) deliveryOutcome {
	var lastErr error
	for attemptIndex := 0; attemptIndex < maxDeliveryAttempts; attemptIndex++ {
		if waitBefore := deliveryRetryBackoff[attemptIndex]; waitBefore > 0 {
			select {
			case <-ctx.Done():
				// Junior Tip [cancelamento não é sucesso]: devolver nil aqui faria o
				// chamador acreditar que gravou. O chunk tem de ir para a fila.
				return deliveryOutcome{err: ctx.Err(), attemptsMade: attemptIndex, terminal: false}
			case <-time.After(waitBefore):
			}
		}
		lastErr = attemptDelivery()
		if lastErr == nil {
			return deliveryOutcome{attemptsMade: attemptIndex + 1}
		}
		if isTerminalDeliveryError(lastErr) {
			return deliveryOutcome{err: lastErr, attemptsMade: attemptIndex + 1, terminal: true}
		}
	}
	return deliveryOutcome{err: lastErr, attemptsMade: maxDeliveryAttempts}
}

// terminalityLabel dá ao log a palavra que faltava.
//
// Junior Tip [a distinção que custou 10 dias]: no log antigo, "flush still
// failing" cobria tanto o banco fora do ar quanto uma rejeição permanente. São
// situações opostas — uma se resolve sozinha, a outra nunca — e quem lia o log não
// tinha como separá-las. Um HTTP 409 ficou 10 dias parecendo uma queda temporária.
func terminalityLabel(terminal bool) string {
	if terminal {
		return "server refused; retrying would not help"
	}
	return "no usable response; will retry from the queue"
}

// isTerminalDeliveryError responde: o servidor já decidiu, ou ainda pode mudar?
//
// Junior Tip [classificar por TIPO, não por texto]: o SDK devolve erros tipados
// (ErrUnauthorized, ErrNotFound, *APIError com StatusCode). Casar por substring na
// mensagem — "401", "unauthorized" — quebraria calado no dia em que alguém
// reescrevesse uma mensagem, e o custo do engano é assimétrico: classificar um
// transiente como terminal desiste cedo demais e cria estado local sem
// necessidade. Na dúvida, o padrão é TENTAR DE NOVO.
func isTerminalDeliveryError(deliveryErr error) bool {
	if deliveryErr == nil {
		return false
	}
	// Chave errada, sem permissão, rota inexistente: repetir dá o mesmo resultado.
	if errors.Is(deliveryErr, client.ErrUnauthorized) || errors.Is(deliveryErr, client.ErrNotFound) {
		return true
	}
	var apiErr *client.APIError
	if errors.As(deliveryErr, &apiErr) {
		switch apiErr.StatusCode {
		case 408, 429:
			// Tempo esgotado e excesso de requisições melhoram sozinhos: transientes.
			return false
		}
		// 4xx = o servidor entendeu e recusou (400 malformado, 409 sessão cheia,
		// 422 inválido). 5xx = o servidor tropeçou e pode voltar: transiente.
		return apiErr.StatusCode >= 400 && apiErr.StatusCode < 500
	}
	// Rede, DNS, timeout, EOF, contexto: tudo isso é "não respondeu".
	return false
}

File: v2/plugins/hermes-agent/delivery_retry.py
"""Bounded retry before anything is written locally.

Junior Tip [a regra do dono, 2026-07-31]: "só pode gravar local caso o banco não
responda depois de 3 tentativas; assim que voltar, descarregar tudo que tem local
e limpar o local pra não ter risco". Antes dela, UMA falha qualquer — um blip de
rede, uma eleição de líder de 200 ms — já criava estado local. Estado local é uma
segunda cópia da verdade: enquanto existe, a memória do usuário está em dois
lugares e só um deles é o AnhurDB. A fila tem de ser um túnel, não um armário.

Junior Tip [3 tentativas, mas NÃO 3 tentativas em tudo]: o dono disse "o banco não
responde". Não responder é uma coisa; responder "não" é outra. Um 401 (chave
errada) ou um 409 (sessão cheia) são o servidor RESPONDENDO, e a resposta não muda
na segunda tentativa — repetir só gastaria o tempo do turno do usuário para chegar
ao mesmo lugar. Então 3 é o TETO para o que pode melhorar sozinho, e o que já foi
decidido vai direto para a fila. O turno é enfileirado nos dois casos: a regra muda
QUANDO se grava local, nunca SE a memória é preservada.

Este módulo é o par exato de ``plugins/core/delivery_retry.go``. Os dois plugins
Go compartilham aquele arquivo byte a byte; este mantém o Hermes Python na mesma
semântica. Mudou um, muda o outro — ver o invariante de paridade dos SDKs.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass
from typing import Callable, Optional, Sequence

# O número do dono: três tentativas antes de aceitar o disco.
MAX_DELIVERY_ATTEMPTS = 3

# Espera ANTES de cada tentativa (a primeira é imediata).
#
# Junior Tip [por que tão curto]: isto roda dentro do turno do usuário. O alvo é o
# blip — eleição de líder, reconexão, um 503 de deploy —, não uma queda longa.
# Queda longa é justamente o que a fila resolve, e a fila não tem pressa.
DELIVERY_RETRY_BACKOFF_SECONDS: Sequence[float] = (0.0, 0.3, 1.2)

# Junior Tip [por que casar por texto AQUI e por tipo no Go]: o SDK Go devolve
# erros tipados (ErrUnauthorized, *APIError com StatusCode) e lá a classificação é
# por tipo. O runtime Python embrulha tudo em AnhurRuntimeError com a mensagem
# original dentro, então o status só existe como texto. Casar por número isolado
# ("409") daria falso positivo em qualquer id que contivesse 409; o padrão abaixo
# exige a palavra "status"/"HTTP" perto do código, ou a frase da API.
_TERMINAL_STATUS_PATTERN = re.compile(
    r"(?:\bhttp\b|\bstatus\b|\bcode\b)\D{0,4}(400|401|403|404|405|409|413|415|422)\b",
    re.IGNORECASE,
)
_TERMINAL_PHRASES = (
    "unauthorized",
    "forbidden",
    "invalid api key",
    "api key is invalid",
    "not found",
    "session has reached the maximum",
    "quota exceeded",
    "validation error",
)


@dataclass
class DeliveryOutcome:
    """Como uma entrega terminou.

    Junior Tip [a distinção que custou 10 dias]: no log antigo, "still failing"
    cobria tanto o banco fora do ar quanto uma rejeição permanente. São situações
    opostas — uma se resolve sozinha, a outra nunca — e quem lia o log não tinha
    como separá-las. Um HTTP 409 passou 10 dias parecendo queda temporária.
    """

    error: Optional[BaseException] = None
    attempts_made: int = 0
    terminal: bool = False

    @property
    def delivered(self) -> bool:
        return self.error is None

    @property
    def terminality_label(self) -> str:
        if self.terminal:
            return "server refused; retrying would not help"
        return "no usable response; will retry from the queue"


def deliver_with_retry(
    attempt_delivery: Callable[[], None],
    *,
    sleep: Callable[[float], None] = time.sleep,
) -> DeliveryOutcome:
    """Executa ``attempt_delivery`` até MAX_DELIVERY_ATTEMPTS vezes.

    Para cedo quando o erro é terminal. ``sleep`` é injetável para os testes
    exercitarem a lógica sem pagar a espera real.
    """
    last_error: Optional[BaseException] = None
    for attempt_index in range(MAX_DELIVERY_ATTEMPTS):
        wait_before = DELIVERY_RETRY_BACKOFF_SECONDS[attempt_index]
        if wait_before > 0:
            sleep(wait_before)
        try:
            attempt_delivery()
            return DeliveryOutcome(attempts_made=attempt_index + 1)
        except Exception as delivery_error:  # noqa: BLE001 — classificado, não engolido
            last_error = delivery_error
            if is_terminal_delivery_error(delivery_error):
                return DeliveryOutcome(
                    error=delivery_error,
                    attempts_made=attempt_index + 1,
                    terminal=True,
                )
    return DeliveryOutcome(error=last_error, attempts_made=MAX_DELIVERY_ATTEMPTS)


def is_terminal_delivery_error(delivery_error: BaseException) -> bool:
    """O servidor já decidiu, ou ainda pode mudar?

    Junior Tip [na dúvida, TENTAR DE NOVO]: o custo do engano é assimétrico.
    Classificar um transiente como terminal desiste cedo e cria estado local sem
    necessidade — exatamente o que a regra do dono existe para evitar. Classificar
    um terminal como transiente custa duas chamadas rápidas a mais. Por isso o
    padrão é falso: só é terminal o que a gente RECONHECE como recusa.
    """
    message = str(delivery_error).lower()
    # 408 (timeout) e 429 (excesso) são 4xx que melhoram sozinhos: não entram na
    # lista acima, de propósito.
    if _TERMINAL_STATUS_PATTERN.search(message):
        return True
    return any(phrase in message for phrase in _TERMINAL_PHRASES)
